File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/WrongViewIdFormatDetectorTest.java
package com.vanniktech.lintrules.android;

import com.android.tools.lint.checks.infrastructure.LintDetectorTest;
import com.android.tools.lint.detector.api.Detector;
import com.android.tools.lint.detector.api.Issue;
import java.util.List;
import org.intellij.lang.annotations.Language;

import static com.vanniktech.lintrules.android.AndroidDetectorTest.NO_WARNINGS;
import static com.vanniktech.lintrules.android.WrongViewIdFormatDetector.ISSUE_WRONG_VIEW_ID_FORMAT;
import static java.util.Collections.singletonList;
import static org.fest.assertions.api.Assertions.assertThat;

public class WrongViewIdFormatDetectorTest extends LintDetectorTest {
  public void testIdLowerCamelCase() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<TextView xmlns:android=\"http://schemas.android.com/apk/res/android\" android:id=\"@+id/lowerCamelCase\"/>\n";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo(NO_WARNINGS);
  }

  public void testIdCamelCase() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<TextView\n"
        + "    xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
        + "    android:id=\"@+id/CamelCase\"/>\n";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo("res/layout/ids.xml:4: Warning: Id is not in lowerCamelCaseFormat [WrongViewIdFormat]\n"
        + "    android:id=\"@+id/CamelCase\"/>\n"
        + "                ~~~~~~~~~~~~~~\n"
        + "0 errors, 1 warnings\n");
  }

  public void testIdSnakeCase() throws Exception {
    @Language("XML") final String source = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n"
        + "<TextView\n"
        + "    xmlns:android=\"http://schemas.android.com/apk/res/android\"\n"
        + "    android:id=\"@+id/snake_case\"/>\n";

    assertThat(lintProject(xml("/res/layout/ids.xml", source))).isEqualTo("res/layout/ids.xml:4: Warning: Id is not in lowerCamelCaseFormat [WrongViewIdFormat]\n"
        + "    android:id=\"@+id/snake_case\"/>\n"
        + "                ~~~~~~~~~~~~~~~\n"
        + "0 errors, 1 warnings\n");
  }

  @Override protected Detector getDetector() {
    return new WrongViewIdFormatDetector();
  }

  @Override protected List<Issue> getIssues() {
    return singletonList(ISSUE_WRONG_VIEW_ID_FORMAT);
  }

  @Override protected boolean allowCompilationErrors() {
    return false;
  }
}

File: lint-rules-android-lint/src/test/java/com/vanniktech/lintrules/android/WrongViewIdFormatDetectorTest.kt
package com.vanniktech.lintrules.android

import com.android.tools.lint.checks.infrastructure.TestFiles.xml
import com.android.tools.lint.checks.infrastructure.TestLintTask.lint
import org.junit.Test

class WrongViewIdFormatDetectorTest {
  @Test fun idLowerCamelCase() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView xmlns:android="http://schemas.android.com/apk/res/android"
          |    android:id="@+id/lowerCamelCase"/>""".trimMargin()))
        .issues(ISSUE_WRONG_VIEW_ID_FORMAT)
        .run()
        .expectClean()
  }

  @Test fun idDefinedLowerCamelCase() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView xmlns:android="http://schemas.android.com/apk/res/android"
          |    android:id="@id/lowerCamelCase"/>""".trimMargin()))
        .issues(ISSUE_WRONG_VIEW_ID_FORMAT)
        .run()
        .expectClean()
  }

  @Test fun idCamelCase() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView xmlns:android="http://schemas.android.com/apk/res/android"
          |    android:id="@+id/CamelCase"/>""".trimMargin()))
        .issues(ISSUE_WRONG_VIEW_ID_FORMAT)
        .run()
        .expect("""
          |res/layout/ids.xml:2: Warning: Id is not in lowerCamelCaseFormat [WrongViewIdFormat]
          |    android:id="@+id/CamelCase"/>
          |                ~~~~~~~~~~~~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/layout/ids.xml line 1: Convert to lowerCamelCase:
          |@@ -2 +2
          |-     android:id="@+id/CamelCase"/>
          |@@ -3 +2
          |+     android:id="@+id/camelCase"/>
          |""".trimMargin())
  }

  @Test fun idSnakeCase() {
    lint()
        .files(xml("res/layout/ids.xml", """
          |<TextView xmlns:android="http://schemas.android.com/apk/res/android"
          |    android:id="@+id/snake_case"/>""".trimMargin()))
        .issues(ISSUE_WRONG_VIEW_ID_FORMAT)
        .run()
        .expect("""
          |res/layout/ids.xml:2: Warning: Id is not in lowerCamelCaseFormat [WrongViewIdFormat]
          |    android:id="@+id/snake_case"/>
          |                ~~~~~~~~~~~~~~~
          |0 errors, 1 warnings""".trimMargin())
        .expectFixDiffs("""
          |Fix for res/layout/ids.xml line 1: Convert to lowerCamelCase:
          |@@ -2 +2
          |-     android:id="@+id/snake_case"/>
          |@@ -3 +2
          |+     android:id="@+id/snakeCase"/>
          |""".trimMargin())
  }
}
